### PR TITLE
fix: heartbeat reconnect

### PR DESCRIPTION
**Pull request by jajeffries**
This pull request refactors the heartbeat management logic in the fleet agent to improve reliability, testability, and concurrency handling. The core change is the replacement of the old ticker-based heartbeat loop with a session-based model that ensures only one heartbeat loop runs at a time, and that it is properly cleaned up on reconnects or shutdowns. The tests have been updated and expanded to cover the new behavior, including edge cases like rapid reconnects and concurrent cancellation.

**Heartbeat session management improvements:**

* Replaced the old `sendHeartbeats` method and per-instance ticker with a new `StartHeartbeats` method in the `heartbeater` struct. This method ensures that only one heartbeat loop runs at a time by canceling and waiting for the previous session before starting a new one. The heartbeat loop now uses a context and a per-session ticker, improving cleanup and preventing concurrent loops. (`agent/configmgr/fleet/heartbeats.go`, `agent/configmgr/fleet/connection.go`) [[1]](diffhunk://#diff-796a5a43e9dd714c20b0d1863be358a6f2b4f3d5035e9ab9c4303f61fb2f24f2R20-R101) [[2]](diffhunk://#diff-796a5a43e9dd714c20b0d1863be358a6f2b4f3d5035e9ab9c4303f61fb2f24f2L137-L161) [[3]](diffhunk://#diff-1783db06b73b72c164682aec902cffc352ef29c1eee8cd92929e7637e72f8047L184-R184)

* Added synchronization primitives (`sync.Mutex`, `sync.WaitGroup`) to the `heartbeater` struct to safely manage concurrent access and ensure proper shutdown of heartbeat goroutines. (`agent/configmgr/fleet/heartbeats.go`)

**Testing improvements and coverage:**

* Updated all heartbeat-related tests to use the new session-based API, removing direct reliance on the ticker and ensuring proper cleanup by waiting for goroutines to finish. This includes adapting expectations to the new publish function signatures and session semantics. (`agent/configmgr/fleet/heartbeats_test.go`) [[1]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L122-L134) [[2]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L214-R228) [[3]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L246-R257) [[4]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L275-R290) [[5]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L317) [[6]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L328-R322) [[7]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L349-R344) [[8]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L379-L387) [[9]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L404-R392)

* Added new tests to verify that multiple stop/start cycles do not result in duplicate or missed heartbeats, and that at most one publish happens at a time, ensuring thread safety and correct session handoff. (`agent/configmgr/fleet/heartbeat_reconnect_test.go`)

**Test configuration and cleanup:**

* Centralized the heartbeat tick interval for tests, allowing for faster test execution and easier configuration. (`agent/configmgr/fleet/heartbeats_test.go`)

* Removed unnecessary ticker initialization and cleanup from the `heartbeater` struct and related test helpers, as the ticker is now managed per session. (`agent/configmgr/fleet/heartbeats.go`, `agent/configmgr/fleet/heartbeats_test.go`) [[1]](diffhunk://#diff-796a5a43e9dd714c20b0d1863be358a6f2b4f3d5035e9ab9c4303f61fb2f24f2R20-R101) [[2]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L79) [[3]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L94) [[4]](diffhunk://#diff-61bcdb2570d98db360b12b22aede9c54e0c453085acd5e60d5902c8f92380e72L107)

These changes make the heartbeat mechanism more robust, prevent subtle bugs related to concurrent sessions, and improve test reliability and maintainability.

### CODE REVIEW

**Review by MicahParks — APPROVED** [review]

File: agent/configmgr/fleet/heartbeats.go [hunk 1]
```diff
@@ -16,29 +17,88 @@ const (
...
+// StartHeartbeats begins a heartbeat session: it cancels any prior session, waits for it to exit,
+// then starts a new loop with its own ticker until the session context is cancelled or stop() runs.
+func (hb *heartbeater) StartHeartbeats(parentCtx context.Context, heartbeatTopic string, agentID string, publishFunc func(ctx context.Context, topic string, payload []byte) error, onFailure func()) {
+	hb.mu.Lock()
+	if hb.sessionCancel != nil {
+		prev := hb.sessionCancel
+		hb.sessionCancel = nil
+		hb.mu.Unlock()
+		prev()
+		hb.wg.Wait()
```
[the remark this answers is not in the document]
Comment (jajeffries) [reply agent/configmgr/fleet/heartbeats.go:56]: **Addressed in 25406e1:** Removed the premature `sessionCancel = nil` before `wg.Wait()` in `StartHeartbeats`. The stored cancel stays live until the prior goroutine has drained, so concurrent `stop()` still sees an active session and invokes the same cancel path instead of treating the heartbeater as idle.

File: agent/configmgr/fleet/connection.go [hunk 2]
```diff
@@ -181,7 +181,7 @@ func (connection *MQTTConnection) Connect(ctx context.Context, details Connectio
 			}
 
 			// start heartbeat loop bound to the same connection-level context
-			go connection.heartbeater.sendHeartbeats(ctx, func() {}, details.Topics.Heartbeat, details.ClientID, connection.publishToTopic, func() {
+			connection.heartbeater.StartHeartbeats(ctx, details.Topics.Heartbeat, details.ClientID, connection.publishToTopic, func() {
```
[the remark this answers is not in the document]
Comment (jajeffries) [reply agent/configmgr/fleet/connection.go]: **Addressed in 25406e1:** Restored `go connection.heartbeater.StartHeartbeats(...)` so `OnConnectionUp` does not block while the prior heartbeat session is cancelled and drained (subscribe/capabilities/group-membership work can proceed).
